src/traits/impls.rs

```diff
@@ -0,0 +1,132 @@
+//!
+//! Doughnut trait impls
+//!
+use super::{DoughnutApi, DoughnutVerify};
+use crate::error::ValidationError;
+#[cfg(feature = "std")]
+use ed25519_dalek::{PublicKey as Ed25519Pub, Signature as Ed25519Sig};
+#[cfg(feature = "std")]
+use schnorrkel::{signing_context, PublicKey as Sr25519Pub, Signature as Sr25519Sig};
+
+#[cfg(feature = "std")]
+impl<T: DoughnutApi> DoughnutVerify for T
+where
+    <T as DoughnutApi>::PublicKey: AsRef<[u8]>,
+    <T as DoughnutApi>::Signature: AsRef<[u8]>,
+{
+    fn verify(&self) -> bool {
+        match self.signature_version() {
+            0 => {
+                let signature = match Sr25519Sig::from_bytes(self.signature().as_ref()) {
+                    Ok(s) => s,
+                    Err(_) => return false,
+                };
+                match Sr25519Pub::from_bytes(self.issuer().as_ref()) {
+                    Ok(pk) => {
+                        // TODO: Parameterise signing context. Hard-coded for now
+                        pk.verify(
+                            signing_context(b"substrate").bytes(self.payload().as_ref()),
+                            &signature,
+                        )
+                    }
+                    Err(_) => return false,
+                }
+            }
+            1 => {
+                let signature = match Ed25519Sig::from_bytes(self.signature().as_ref()) {
+                    Ok(s) => s,
+                    Err(_) => return false,
+                };
+                match Ed25519Pub::from_bytes(self.issuer().as_ref()) {
+                    Ok(pk) => match pk.verify(&self.payload(), &signature) {
+                        Ok(_) => true,
+                        _ => false,
+                    },
+                    Err(_) => false,
+                }
+            }
+            _ => {
+                // Unsupported
+                false
+            }
+        }
+    }
+}
+
+// Dummy implementation for unit type
+impl DoughnutApi for () {
+    type PublicKey = ();
+    type Timestamp = ();
+    type Signature = ();
+    fn holder(&self) -> Self::PublicKey {
+        ()
+    }
+    fn issuer(&self) -> Self::PublicKey {
+        ()
+    }
+    fn expiry(&self) -> Self::PublicKey {
+        ()
+    }
+    fn payload(&self) -> Vec<u8> {
+        Default::default()
+    }
+    fn signature(&self) -> Self::Signature {
+        ()
+    }
+    fn signature_version(&self) -> u8 {
+        255
+    }
+    fn get_domain(&self, _domain: &str) -> Option<&[u8]> {
```

Comment:
Does it make sense to make this capable of extracting multiple domains at once? Lookup is expensive each time, given it needs to recurse, if we're not caching the decoded domain index.

Reply:
That would be good but my main concern is waiting until this proves to be an issue in practice, especially since the actual usage is still mostly theory

Reply:
It's less theoretical than you'd think, in terms of business cases waiting for the technology. But let's add it to the backlog to be reviewed again at a later date.